File: src/stores/library.ts
import { create } from 'zustand';
import { invokeBackend } from '../api/backend';

const toMessage = (error: unknown): string =>
  error instanceof Error ? error.message : String(error);

export interface Track {
  id: string;
  title: string;
  artist: string;
  album: string;
  duration: number;
  file_path: string;
  format: string;
}

export interface LibraryStats {
  total_tracks: number;
  total_albums: number;
  total_artists: number;
}

interface RawTrack {
  track_id: string;
  id?: string;
  title: string;
  artist: string | null;
  album: string | null;
  duration: number | null;
  duration_ms: number | null;
  local_file_path: string | null;
  file_path?: string | null;
  format?: string;
}

const normalizeTrack = (raw: RawTrack): Track => ({
  id: raw.track_id ?? raw.id ?? '',
  title: raw.title,
  artist: raw.artist ?? 'Unknown Artist',
  album: raw.album ?? '',
  duration: raw.duration ?? (raw.duration_ms ? raw.duration_ms / 1000 : 0),
  file_path: raw.local_file_path ?? raw.file_path ?? '',
  format: raw.format ?? '',
});

interface LibraryStore {
  tracks: Track[];
  stats: LibraryStats | null;
  favorites: string[];
  isLoading: boolean;
  error: string | null;
  fetchLibrary: () => Promise<void>;
  fetchFavorites: () => Promise<void>;
  toggleFavorite: (trackId: string) => Promise<void>;
  scanFolder: (folderPath: string) => Promise<void>;
  importFiles: (filePaths: string[]) => Promise<void>;
  importSpotify: (url: string) => Promise<void>;
}

export const useLibraryStore = create<LibraryStore>((set) => ({
  tracks: [],
  stats: null,
  favorites: [],
  isLoading: false,
  error: null,

  fetchLibrary: async () => {
    set({ isLoading: true, error: null });
    try {
      const rawTracks = await invokeBackend('get_library') as unknown as RawTrack[];
      const tracks = rawTracks.map(normalizeTrack);
      set({ tracks, isLoading: false });
    } catch (error) {
      set({ error: toMessage(error), isLoading: false });
    }
  },

  fetchFavorites: async () => {
    try {
      const favorites = await invokeBackend('get_favorites');
      set({ favorites });
    } catch (error) {
      set({ error: toMessage(error) });
    }
  },

  toggleFavorite: async (trackId: string) => {
    try {
      const result = await invokeBackend('toggle_favorite', { track_id: trackId });
      set({ favorites: result.favorites });
    } catch (error) {
      set({ error: toMessage(error) });
    }
  },
  
  scanFolder: async (folderPath: string) => {
    set({ isLoading: true, error: null });
    try {
      await invokeBackend('scan_folder', { folderPath });
      // Refresh library after scan
      const rawTracks = await invokeBackend('get_library') as unknown as RawTrack[];
      const tracks = rawTracks.map(normalizeTrack);
      set({ tracks, isLoading: false });
    } catch (error) {
      set({ error: toMessage(error), isLoading: false });
    }
  },
  
  importFiles: async (filePaths: string[]) => {
    set({ isLoading: true, error: null });
    try {
      await invokeBackend('import_files', { filePaths });
      // Refresh library after import
      const rawTracks = await invokeBackend('get_library') as unknown as RawTrack[];
      const tracks = rawTracks.map(normalizeTrack);
      set({ tracks, isLoading: false });
    } catch (error) {
      set({ error: toMessage(error), isLoading: false });
    }
  },

  importSpotify: async (url: string) => {
    set({ isLoading: true, error: null });
    try {
      await invokeBackend('import_spotify', { url });
      const rawTracks = await invokeBackend('get_library') as unknown as RawTrack[];
      const tracks = rawTracks.map(normalizeTrack);
      set({ tracks, isLoading: false });
    } catch (error) {
      set({ error: toMessage(error), isLoading: false });
    }
  },
}));

File: src-tauri/src/commands/library.rs
use crate::AppState;
use tauri::State;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Track {
    pub id: String,
    pub title: String,
    pub artist: String,
    pub album: String,
    pub duration: u32,
    pub file_path: String,
    pub format: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LibraryStats {
    pub total_tracks: usize,
    pub total_albums: usize,
    pub total_artists: usize,
}

#[tauri::command]
pub async fn get_library(state: State<'_, AppState>) -> Result<Vec<Track>, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let tracks = db.get_all_tracks().map_err(|e| e.to_string())?;
    Ok(tracks)
}

#[tauri::command]
pub async fn get_library_stats(state: State<'_, AppState>) -> Result<LibraryStats, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let stats = db.get_library_stats().map_err(|e| e.to_string())?;
    Ok(stats)
}

#[tauri::command]
pub async fn scan_folder(folder_path: String, state: State<'_, AppState>) -> Result<String, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let scanner = crate::filesystem::scanner::FileScanner::new();
    let tracks = scanner.scan_folder(&folder_path).map_err(|e| e.to_string())?;
    let track_count = tracks.len();
    
    for track in tracks {
        db.add_track(&track).map_err(|e| e.to_string())?;
    }
    
    Ok(format!("Scanned {} tracks from {}", track_count, folder_path))
}

#[tauri::command]
pub async fn import_files(file_paths: Vec<String>, state: State<'_, AppState>) -> Result<String, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let mut imported = 0;
    
    for file_path in file_paths {
        let metadata = crate::audio::metadata::extract_metadata(&file_path).map_err(|e| e.to_string())?;
        db.add_track(&metadata).map_err(|e| e.to_string())?;
        imported += 1;
    }
    
    Ok(format!("Imported {} tracks", imported))
}